README.md

# Nessus Cardano

With Nessus Cardano we explore, praise, comment, contribute to various technical aspects of [Cardano](https://cardano.org).
This is our contribution to "Making The World Work Better For All".

Initially, we focus on a "container first" approach for the Cardano node.

## Running a Node

To get up and running with [Cardano](https://cardano.org), you can spin up a node like this ...

```
docker run --detach \
    --name=relay \
    -p 3001:3001 \
    -v node-data:/opt/cardano/data \
    -v node-ipc:/opt/cardano/ipc \
    nessusio/cardano-node run

docker logs -f relay
```

This works on [x86_64](https://hub.docker.com/r/nessusio/cardano-node/tags?name=amd64) and [arm64](https://hub.docker.com/r/nessusio/cardano-node/tags?name=arm64).

The [nessusio/cardano-node](https://hub.docker.com/r/nessusio/cardano-node) image is built from source in multiple stages like [this](nix/cardano/Dockerfile) and then
with [Nix](https://nixos.org) like [this](nix/docker/node/default.nix).

## Running a Node on the Testnet

```
docker run --detach \
    --name=testrl \
    -p 3001:3001 \
    -e CARDANO_NETWORK=testnet \
    -v test-data:/opt/cardano/data \
    -v node-ipc:/opt/cardano/ipc \
    nessusio/cardano-node run

docker logs -f testrl
```

## Accessing the build-in gLiveView

The image has [gLiveView](https://github.com/cardano-community/guild-operators/blob/alpha/scripts/cnode-helper-scripts/gLiveView.sh) monitoring built in.

For a running container, you can do ...

```
docker exec -it relay gLiveView
```

<img src="docs/img/relay-glview.png" width="500">

## Accessing the build-in topology updater

There is currently no P2P module activated in cardano-1.26.1. Your node may call out to well known relay nodes, but you may never have incoming connections.
According to [this](https://github.com/cardano-community/guild-operators/blob/alpha/docs/Scripts/topologyupdater.md) it is necessary to update your topology
every hour. At the time of writing, cardano-node doesn't do this on its own.

This functionality has been built into nessus/cardano-node as well. On startup, you should see a log similar to this ...

```
docker run --detach \
    --name=relay \
    -p 3001:3001 \
    -e CARDANO_UPDATE_TOPOLOGY=true \
    -v node-data:/opt/cardano/data \
    nessusio/cardano-node run

$ docker exec -it relay tail /opt/cardano/logs/topologyUpdateResult
{ "resultcode": "201", "datetime":"2021-01-10 18:30:06", "clientIp": "209.250.233.200", "iptype": 4, "msg": "nice to meet you" }
{ "resultcode": "203", "datetime":"2021-01-10 19:30:03", "clientIp": "209.250.233.200", "iptype": 4, "msg": "welcome to the topology" }
{ "resultcode": "204", "datetime":"2021-01-10 20:30:04", "clientIp": "209.250.233.200", "iptype": 4, "msg": "glad you're staying with us" }
```

The topologyUpdater is triggered by `CARDANO_UPDATE_TOPOLOGY`. Without it, the cron job is not installed.

## External storage for block data

In this configuration, we map the node's `--database-path` to a mounted directory.

```
docker run --detach \
    --name=relay \
    -p 3001:3001 \
    -v /mnt/disks/data00:/opt/cardano/data \
    nessusio/cardano-node run

docker logs -f relay
```

## Using custom configurations

Here we define a config volume called `cardano-relay-config`. It holds the `mainnet-topology.json` file that we setup externally.
Note, that our custom config lives in `/var/cardano/config` and not in the default location `/opt/cardano/config`.

```
docker run --detach \
    --name=relay \
    -p 3001:3001 \
    -e CARDANO_UPDATE_TOPOLOGY=true \
    -e CARDANO_CUSTOM_PEERS="$PRODUCER_IP:3001" \
    -e CARDANO_TOPOLOGY="/var/cardano/config/mainnet-topology.json" \
    -v cardano-relay-config:/var/cardano/config  \
    -v /mnt/disks/data00:/opt/cardano/data \
    nessusio/cardano-node run

docker logs -f relay
```

## Running a Block Producer Node

A producer node is configured in the same way as a Relay node, except that it has some additional key/cert files configured and does not need to update its topology.

```
docker run --detach \
    --name=bprod \
    -p 3001:3001 \
    -e CARDANO_BLOCK_PRODUCER=true \
    -e CARDANO_TOPOLOGY="/var/cardano/config/mainnet-topology.json" \
    -e CARDANO_SHELLEY_KES_KEY="/var/cardano/config/keys/pool/kes.skey" \
    -e CARDANO_SHELLEY_VRF_KEY="/var/cardano/config/keys/pool/vrf.skey" \
    -e CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE="/var/cardano/config/keys/pool/node.cert" \
    -v cardano-prod-config:/var/cardano/config  \
    -v /mnt/disks/data01:/opt/cardano/data \
    nessusio/cardano-node run

docker logs -f bprod

docker exec -it bprod gLiveView
```

## Running the Cardano CLI

We can also use the image to run Cardano CLI commands.

For this to work, the node must share its IPC socket location, which can then
be use in the alias definition.

```
alias cardano-cli="docker run -it --rm \
  -v ~/cardano:/var/cardano/local \
  -v node-ipc:/opt/cardano/ipc \
  nessusio/cardano-node cardano-cli"

cardano-cli query tip --mainnet
{
    "blockNo": 5102089,
    "headerHash": "e5984f27d1d3b5dcc296b33ccd919a28618ff2d77513971bd316cffd35afecda",
    "slotNo": 16910651
}
```

## Docker Compose

We sometimes may prefer somm middle ground between manually spinning up individual docker containers and the full blown enterprise Kubernetes account.

Perhaps we'd like to use [Docker Compose](https://docs.docker.com/compose).

```
$ docker-compose -f nix/docker/compose/cardano-node-relay.yaml up --detach

Creating relay ... done
Creating nginx ... done

$ docker-compose -f nix/docker/compose/cardano-node-bprod.yaml up --detach

Creating bprod ... done
Creating nginx ... done
```

For details you may want to have a look at [nix/docker/compose/cardano-nodes.yaml](https://github.com/tdiesler/nessus-cardano/blob/master/nix/docker/compose/cardano-nodes.yaml).

## Kubernetes

This project has started as an incubator space for stuff that may eventually become available upstream. As part of this, we want to
"[provide high quality multiarch docker images and k8s support](https://forum.cardano.org/t/provide-high-quality-multiarch-docker-image-and-k8s-support/47906)"

With Kubernetes as the de-facto standard for container deployment, orchestration, monitoring, scaling , etc, it should be as easy as this to run Cardano nodes …


```
kubectl apply -f nix/docker/k8s/cardano-nodes.yaml

storageclass.storage.k8s.io/cardano-standard-rwo created
persistentvolumeclaim/relay-data created
pod/relay created
service/relay-np created
service/relay-clip created
persistentvolumeclaim/bprod-data created
pod/bprod created
service/bprod-clip created
```

This is documented in detail [over here](https://app.gitbook.com/@tdiesler/s/cardano/k8s/cardano-k8s).

For details you may want to have a look at [nix/docker/k8s/cardano-nodes.yaml](https://github.com/tdiesler/nessus-cardano/blob/master/nix/docker/k8s/cardano-nodes.yaml).

## Leader logs

For a Stake Pool Operator it is important to know when the node is scheduled to produce the next block. We definitely want to be online at that important moment and fullfil our block producing duties. There are better times to do node maintenance.

This important functionality has also been built into the [nessusio/cardano-tools](https://hub.docker.com/r/nessusio/cardano-tools) image.

First, lets define an alias and ping the node that we want to work with.

Details about this API are [here](https://github.com/AndrewWestberg/cncli/blob/develop/USAGE.md).

```
$ alias cncli="docker run -it --rm \
  -v ~/cardano:/var/cardano/local \
  -v cncli:/var/cardano/cncli \
  nessusio/cardano-tools cncli"

NODE_IP=192.168.0.30

cncli ping --host $NODE_IP
{
  "status": "ok",
  "host": "10.128.0.31",
  "port": 3001,
  "connectDurationMs": 0,
  "durationMs": 53
}
```

### Syncing the database

This command connects to a remote node and synchronizes blocks to a local sqlite database.

```
$ cncli sync --host $NODE_IP \
  --db /var/cardano/cncli/cncli.db \
  --no-service

...
2021-03-04T10:23:19.719Z INFO  cardano_ouroboros_network::protocols::chainsync   > block 5417518 of 5417518, 100.00% synced
2021-03-04T10:23:23.459Z INFO  cncli::nodeclient::sync                           > Exiting...
```

### Slot Leader Schedule

We can now obtain the leader schedule for our pool.

```
STAKE_SNAPSHOT=$HOME/cardano/scratch/stake-snapshot.json
cardano-cli query stake-snapshot --stake-pool-id 9e8009b249142d80144dfb681984e08d96d51c2085e8bb6d9d1831d2 --mainnet | \
  tee $STAKE_SNAPSHOT

# Prev
LEDGER_SET=prev
POOL_STAKE=$(cat $STAKE_SNAPSHOT | jq .poolStakeGo)
ACTIVE_STAKE=$(cat $STAKE_SNAPSHOT | jq .activeStakeGo)

# Current
LEDGER_SET=current
POOL_STAKE=$(cat $STAKE_SNAPSHOT | jq .poolStakeSet)
ACTIVE_STAKE=$(cat $STAKE_SNAPSHOT | jq .activeStakeSet)

# Next
LEDGER_SET=next
POOL_STAKE=$(cat $STAKE_SNAPSHOT | jq .poolStakeMark)
ACTIVE_STAKE=$(cat $STAKE_SNAPSHOT | jq .activeStakeMark)

$ cncli leaderlog \
  --pool-id 9e8009b249142d80144dfb681984e08d96d51c2085e8bb6d9d1831d2 \
  --shelley-genesis /opt/cardano/config/mainnet-shelley-genesis.json \
  --byron-genesis /opt/cardano/config/mainnet-byron-genesis.json \
  --pool-vrf-skey /var/cardano/local/mainnet/keys/pool/vrf.skey \
  --db /var/cardano/cncli/cncli.db \
  --active-stake $ACTIVE_STAKE \
  --pool-stake $POOL_STAKE \
  --tz Europe/Berlin \
  --ledger-set $LEDGER_SET | tee leaderlog.json

cat leaderlog.json | jq -c ".assignedSlots[] | {no: .no, slot: .slotInEpoch, at: .at}"

{"no":1,"slot":165351,"at":"2021-02-26T20:40:42+01:00"}
{"no":2,"slot":312656,"at":"2021-02-28T13:35:47+01:00"}
{"no":3,"slot":330588,"at":"2021-02-28T18:34:39+01:00"}
{"no":4,"slot":401912,"at":"2021-03-01T14:23:23+01:00"}
```

# Build the Images

In line with the [upstream project](https://github.com/input-output-hk/cardano-node) we also us
a [Nix](https://nixos.org/guides/install-nix.html) based build. The build requires Nix and a
working Docker environment. This works on x86_64 and arm64.

Spinning up a build/test environment on GCE is documented [here](./vps/gce-centos8.md)

To build all images and their respective dependencies, run ...

```
./build.sh all
```

Enjoy!


ChangeLog.md

# v1.34.1-rev1

Released: 09-Mar-2022

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.34.1 ([#114][114])
* Upgrade to gLiveView-1.26.3 ([#116][116])
* Add support for cardano-submit-api ([#115][115])
* Frequent Tx timeout in tokenswap proxy ([#112][112])
* Multiply Astor token denomination with 100 ([#111][111])

[111]: https://github.com/tdiesler/nessus-cardano/issues/111
[112]: https://github.com/tdiesler/nessus-cardano/issues/112
[114]: https://github.com/tdiesler/nessus-cardano/issues/114
[115]: https://github.com/tdiesler/nessus-cardano/issues/115
[116]: https://github.com/tdiesler/nessus-cardano/issues/116

For details see [v1.34.1-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.34.1-rev1)

# v1.33.0-rev1

Released: 07-Jan-2021

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.33.0 task ([#108][108])
* Upgrade to gLiveView-1.25.1 task ([#109][109])
* Remove monit from docker-compose scripts ([#105][105])

[105]: https://github.com/tdiesler/nessus-cardano/issues/105
[108]: https://github.com/tdiesler/nessus-cardano/issues/108
[109]: https://github.com/tdiesler/nessus-cardano/issues/109

For details see [v1.33.0-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.33.0-rev1)

# v1.32.1-rev1

Released: 14-Dec-2021

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.32.1 task ([#100][100])
* Upgrade to gLiveView-1.25.0 task ([#101][101])
* Upgrade to cncli-4.0.4 task ([#102][102])
* Add support for astor script --try-run ([#99][99])

[99]: https://github.com/tdiesler/nessus-cardano/issues/99
[100]: https://github.com/tdiesler/nessus-cardano/issues/100
[101]: https://github.com/tdiesler/nessus-cardano/issues/101
[102]: https://github.com/tdiesler/nessus-cardano/issues/102

For details see [v1.32.1-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.32.1-rev1)

# v1.31.0-rev1

Released: 28-Nov-2021

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.31.0 task ([#96][96])
* Upgrade to gLiveView-1.24.0 task ([#97][97])
* Upgrade to cncli-4.0.3 task ([#98][98])
* Add support for leaderlog on arm64 ([#82][82])
* Add script to restart the node at defined times feature ([#89][89])
* Add CARDANO_MAX_PEERS to the configuration feature ([#90][90])
* Make CARDANO_BLOCK_PRODUCER configurable externally feature ([#95][95])
* Cannot use testnet addresses with astorPay2PKH bug ([#94][94])
* Add plutus tokenswap smart contract feature ([#85][85])
* Monitor pool tickers for unreasonable updates feature ([#84][84])

[82]: https://github.com/tdiesler/nessus-cardano/issues/82
[84]: https://github.com/tdiesler/nessus-cardano/issues/84
[85]: https://github.com/tdiesler/nessus-cardano/issues/85
[89]: https://github.com/tdiesler/nessus-cardano/issues/89
[90]: https://github.com/tdiesler/nessus-cardano/issues/90
[94]: https://github.com/tdiesler/nessus-cardano/issues/94
[95]: https://github.com/tdiesler/nessus-cardano/issues/95
[96]: https://github.com/tdiesler/nessus-cardano/issues/96
[97]: https://github.com/tdiesler/nessus-cardano/issues/97
[98]: https://github.com/tdiesler/nessus-cardano/issues/98

For details see [v1.31.0-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.31.0-rev1)

# v1.30.1-rev1

Released: 29-Sep-2021

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.30.1 ([#78][78])
* Upgrade to cardano-db-sync-11.0.0 ([#75][75])
* Upgrade to gLiveView-1.22.3 ([#80][80])
* Upgrade to cncli-4.0.1 ([#79][79])
* Attach the arm64 binaries to the release ([#81][81])

[75]: https://github.com/tdiesler/nessus-cardano/issues/75
[78]: https://github.com/tdiesler/nessus-cardano/issues/78
[79]: https://github.com/tdiesler/nessus-cardano/issues/79
[80]: https://github.com/tdiesler/nessus-cardano/issues/80
[81]: https://github.com/tdiesler/nessus-cardano/issues/81

For details see [v1.30.1-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.30.1-rev1)

# v1.29.0-rev1

Released: 31-Aug-2021

This is a major feature update and bug fix release.

* Upgrade to cardano-node-1.29.0 ([#73][73])
* Additional RTS flags for optimization of memory usage ([#67][67])
* Add support for RTS options in docker run ([#68][68])
* cardano-cli cannot parse script-data-value with spaces ([#69][69])

[67]: https://github.com/tdiesler/nessus-cardano/issues/67
[68]: https://github.com/tdiesler/nessus-cardano/issues/68
[69]: https://github.com/tdiesler/nessus-cardano/issues/69
[73]: https://github.com/tdiesler/nessus-cardano/issues/73

For details see [v1.29.0-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.29.0-rev1)

# v1.28.0-rev1

Released: 21-Jul-2021

This is a major feature update and bug fix release.

* Add support for cncli on arm64 ([#61][61])
* Add support for Log Navigator ([#62][62])
* Upgrade to cardano-node-1.28.0 ([#63][63])
* Upgrade to gLiveView-1.20.10 ([#64][64])
* Upgrade to cncli-3.1.4 ([#65][65])

[61]: https://github.com/tdiesler/nessus-cardano/issues/61
[62]: https://github.com/tdiesler/nessus-cardano/issues/62
[63]: https://github.com/tdiesler/nessus-cardano/issues/63
[64]: https://github.com/tdiesler/nessus-cardano/issues/64
[65]: https://github.com/tdiesler/nessus-cardano/issues/65

For details see [v1.28.0-rev1](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.28.0-rev1)

# v1.27.0-rev3

Released: 15-Jun-2021

This is a minor feature update and bug fix release.

* Simplify stop signal handling by using SIGINT directly ([#49][49])
* Migrate monit to system service ([#60][60])

[49]: https://github.com/tdiesler/nessus-cardano/issues/49
[60]: https://github.com/tdiesler/nessus-cardano/issues/60

For details see [v1.27.0-rev3](https://github.com/tdiesler/nessus-cardano/issues?q=milestone%3Av1.27.0-rev3)

# v1.27.0-rev2

Released: 05-Jun-2021

This is a minor feature update and bug fix release.


node/docker/entrypoint.sh
#!/bin/bash

# Optional ENV vars
#

CARDANO_CONFIG_BASE="/opt/cardano/config"

if [ -v $CARDANO_CONFIG ]; then
  CARDANO_CONFIG="$CARDANO_CONFIG_BASE/mainnet-config.json"
fi

if [ -v $CARDANO_TOPOLOGY ]; then
  CARDANO_TOPOLOGY="$CARDANO_CONFIG_BASE/mainnet-topology.json"
fi

if [ -v $CARDANO_DATABASE_PATH ]; then
  CARDANO_DATABASE_PATH="/opt/cardano/data"
fi

if [ -v $CARDANO_SOCKET_PATH ]; then
  CARDANO_SOCKET_PATH="$CARDANO_DATABASE_PATH/socket"
fi

if [ -v $CARDANO_LOG_DIR ]; then
  CARDANO_LOG_DIR="/opt/cardano/logs"
fi

if [ -v $CARDANO_BIND_ADDR ]; then
  CARDANO_BIND_ADDR="0.0.0.0"
fi

if [ -v $CARDANO_PORT ]; then
  CARDANO_PORT=3001
fi

if [ -v $CARDANO_UPDATE_TOPOLOGY ]; then
  CARDANO_UPDATE_TOPOLOGY=false
fi

if [ -v $CARDANO_BLOCK_PRODUCER ]; then
  CARDANO_BLOCK_PRODUCER=false
fi

# Trap the SIGTERM signal sent from `docker stop` and 
# signal the cardano-node with SIGINT for graceful shutdown
#
# The wait may return before the gracefull shutdown is done
# Wait a little longer before will kill everything

sigtermHandler () {
  echo "Signalling for shutdown ..."
  kill -SIGINT $1
  wait $1
  sleep 2
}

#####################################################################
#
# Print run environment
#
printRunEnv () {

  echo "CARDANO_BIND_ADDR=$CARDANO_BIND_ADDR"
  echo "CARDANO_BLOCK_PRODUCER=$CARDANO_BLOCK_PRODUCER"
  echo "CARDANO_CONFIG=$CARDANO_CONFIG"
  echo "CARDANO_CUSTOM_PEERS=$CARDANO_CUSTOM_PEERS"
  echo "CARDANO_DATABASE_PATH=$CARDANO_DATABASE_PATH"
  echo "CARDANO_LOG_DIR=$CARDANO_LOG_DIR"
  echo "CARDANO_PORT=$CARDANO_PORT"
  echo "CARDANO_PUBLIC_IP=$CARDANO_PUBLIC_IP"
  echo "CARDANO_SOCKET_PATH=$CARDANO_SOCKET_PATH"
  echo "CARDANO_TOPOLOGY=$CARDANO_TOPOLOGY"
  echo "CARDANO_UPDATE_TOPOLOGY=$CARDANO_UPDATE_TOPOLOGY"
  
  if [ "$CARDANO_BLOCK_PRODUCER" = true ]; then
  
    if [ -v $CARDANO_SHELLEY_KES_KEY ]; then
      CARDANO_SHELLEY_KES_KEY="$CARDANO_CONFIG_BASE/keys/kes.skey"
    fi
    
    if [ -v $CARDANO_SHELLEY_VRF_KEY ]; then
      CARDANO_SHELLEY_VRF_KEY="$CARDANO_CONFIG_BASE/keys/vrf.skey"
    fi
    
    if [ -v $CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE ]; then
      CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE="$CARDANO_CONFIG_BASE/keys/node.cert"
    fi
  
    echo "CARDANO_SHELLEY_KES_KEY=$CARDANO_SHELLEY_KES_KEY"
    echo "CARDANO_SHELLEY_VRF_KEY=$CARDANO_SHELLEY_VRF_KEY"
    echo "CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE=$CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE"
  fi
}

#####################################################################
#
# Write root env file
#
writeRootEnv () {
cat << EOF > /root/env
#!/usr/bin/env bash

# Docker run ENV vars 
export CARDANO_BIND_ADDR="$CARDANO_BIND_ADDR"
export CARDANO_BLOCK_PRODUCER="$CARDANO_BLOCK_PRODUCER"
export CARDANO_CONFIG="$CARDANO_CONFIG"
export CARDANO_DATABASE_PATH="$CARDANO_DATABASE_PATH"
export CARDANO_LOG_DIR="$CARDANO_LOG_DIR"
export CARDANO_PORT="$CARDANO_PORT"
export CARDANO_SOCKET_PATH="$CARDANO_SOCKET_PATH"
export CARDANO_TOPOLOGY="$CARDANO_TOPOLOGY"
export CARDANO_UPDATE_TOPOLOGY="$CARDANO_UPDATE_TOPOLOGY"

# Mapping for gLiveView
export CCLI="/usr/local/bin/cardano-cli"
export CONFIG="$CARDANO_CONFIG"
export TOPOLOGY="$CARDANO_TOPOLOGY"
export SOCKET="$CARDANO_SOCKET_PATH"
export LOG_DIR="$CARDANO_LOG_DIR"
export CNODE_PORT="$CARDANO_PORT"

# Mapping for topologyUpdater
export CARDANO_PUBLIC_IP="$CARDANO_PUBLIC_IP"
export CARDANO_CUSTOM_PEERS="$CARDANO_CUSTOM_PEERS"
EOF
}

#####################################################################
#
# Install the topology updater cron job
#
topologyUpdaterCron () {

  
  if [ "$CARDANO_UPDATE_TOPOLOGY" = true ]; then
          
    TARGET="/etc/cron.d/topologyUpdater"
    if [ ! -f "$TARGET" ]; then
      
      MIN=`date +"%M"`
      MOD=$(($MIN % 5))
      MIN=$(($(($MIN - $MOD)) + 10))
      if (( $MIN > 59 )); then MIN=$(($MIN - 60)); fi
    
      CRONJOB="$MIN * * * *  root  /usr/local/bin/topologyUpdater"
      echo "Installing cron: $CRONJOB"
      
      echo "$CRONJOB" > $TARGET
      
    else
    
      CRONJOB="`cat /etc/cron.d/topologyUpdater`"
      echo "Running cron: $CRONJOB"
      
    fi

    service cron start
  fi
}

#####################################################################
#
# Run the relay node in the background
#
runRelayNode () {

  cardano-node run \
    --config $CARDANO_CONFIG \
    --topology $CARDANO_TOPOLOGY \
    --database-path $CARDANO_DATABASE_PATH \
    --socket-path $CARDANO_SOCKET_PATH \
    --host-addr $CARDANO_BIND_ADDR \
    --port $CARDANO_PORT &
}

#####################################################################
#
# Run the block producer in the background
#
runBlockProducerNode () {

  cardano-node run \
    --config $CARDANO_CONFIG \
    --topology $CARDANO_TOPOLOGY \
    --database-path $CARDANO_DATABASE_PATH \
    --socket-path $CARDANO_SOCKET_PATH \
    --host-addr $CARDANO_BIND_ADDR \
    --port $CARDANO_PORT \
    --shelley-kes-key $CARDANO_SHELLEY_KES_KEY \
    --shelley-vrf-key $CARDANO_SHELLEY_VRF_KEY \
    --shelley-operational-certificate $CARDANO_SHELLEY_OPERATIONAL_CERTIFICATE &
}

#####################################################################
#
# Trap the SIGTERM signal
#
trapSIGTERM () {

  PID="$!"
  trap "sigtermHandler $PID" SIGTERM

  # Wait for the child process to terminate
  wait $PID
}

#######################################################################################################################
#
# Command Switch
#
if [ "$1" == "run" ]; then

  printRunEnv
  writeRootEnv

  if [ "$CARDANO_BLOCK_PRODUCER" = true ]; then
      
    runBlockProducerNode
    
  else
  
    topologyUpdaterCron
    
    runRelayNode
  fi
  
  trapSIGTERM
  
elif [ "$1" == "cardano-cli" ]; then

  exec "$@"

elif [ "$1" == "ledger-state" ]; then

  echo "Generating $CARDANO_DATABASE_PATH/ledger-state.json"
  
  cardano-cli query ledger-state \
      --mainnet \
      --allegra-era \
      --out-file $CARDANO_DATABASE_PATH/ledger-state.json

elif [ "$1" == "sigma" ]; then

  CMD="/usr/bin/python3 /root/pooltool.io/leaderLogs/getSigma.py"
  ARGS=`echo $* | cut -d' ' -f2-`
  
  eval "$CMD $ARGS"

elif [ "$1" == "leader-logs" ]; then

  CMD="/usr/bin/python3 /root/pooltool.io/leaderLogs/leaderLogs.py"
  ARGS=`echo $* | cut -d' ' -f2-`
  
  eval "$CMD $ARGS"

else
  
  echo "Nothing to do! Perhaps try [run|cardano-cli|ledger-state|sigma|leader-logs]"
  exit 1
  
fi
